tests/test_overdrive.py

```diff
@@ -276,6 +277,205 @@ def test_scope_string(self):
+        api.PROCESS_CHECKOUT_ERROR_RESULT = "Actually, I was able to recover"
+        api.queue_response(400, content=api_response)
+        eq_(
+            "Actually, I was able to recover",
+            api.checkout(
+                patron, pin, pool, "internal format is ignored"
+            )
+        )
+        eq_((patron, pin, pool, "some data"), api._process_checkout_error_called_with.pop())
+
+    def test__process_checkout_error(self):
+        # Verify that _process_checkout_error handles common API-side errors,
+        # making follow-up API calls if necessary.
+
+        class Mock(MockOverdriveAPI):
+            MOCK_LOAN = object()
+            MOCK_EXPIRATION_DATE = object()
+
+            def __init__(self, *args, **kwargs):
+                super(Mock, self).__init__(*args, **kwargs)
+                self.update_licensepool_called_with = []
+                self.get_loan_called_with = []
+                self.extract_expiration_date_called_with = []
+
+            def update_licensepool(self, identifier):
+                self.update_licensepool_called_with.append(identifier)
+
+            def get_loan(self, patron, pin, identifier):
+                self.get_loan_called_with.append((patron, pin, identifier))
+                return self.MOCK_LOAN
+
+            def extract_expiration_date(self, loan):
+                self.extract_expiration_date_called_with.append(loan)
+                return self.MOCK_EXPIRATION_DATE
+
+        patron = object()
+        pin = object()
+        pool = self._licensepool(edition=None, collection=self.collection)
+        identifier = pool.identifier
+        api = Mock(self._db, self.collection)
+        m = api._process_checkout_error
+
+        # Most of the error handling is pretty straightforward.
+        def with_error_code(code):
+            # Simulate the response of the Overdrive API with a given error code.
+            error = dict(errorCode=code)
+
+            # Handle the error.
+            return m(patron, pin, pool, error)
+
+        # Errors not specifically known become generic CannotLoan exceptions.
+        assert_raises_regexp(
+            CannotLoan, "WeirdError", with_error_code, "WeirdError"
+        )
+
+        # If the data passed in to _process_checkout_error is not what
+        # the real Overdrive API would send, the error is even more
+        # generic.
+        assert_raises_regexp(
+            CannotLoan, "Unknown Error",
+            m, patron, pin, pool, "Not a dict"
+        )
+        assert_raises_regexp(
+            CannotLoan, "Unknown Error",
+            m, patron, pin, pool, dict(errorCodePresent=False)
+        )
+
+        # Some known errors become specific subclasses of CannotLoan.
+        assert_raises(PatronLoanLimitReached, with_error_code,
+                      "PatronHasExceededCheckoutLimit")
+        assert_raises(PatronLoanLimitReached, with_error_code,
+                      "PatronHasExceededCheckoutLimit_ForCPC")
```

Comment:
Here is the test of the new functionality; everything else is coverage that was previously missing.